File: packages/core/test/geo/lambert93.test.ts
import { describe, expect, it } from 'vitest'
import {
  boundsOf,
  boundsContain,
  expandBounds,
  forward,
  forwardPoint,
  inversePoint,
  type LatLon,
} from '../../src/geo/lambert93.js'

/** Reference eastings/northings produced by pyproj (EPSG:4326 -> EPSG:2154). */
const references: ReadonlyArray<readonly [LatLon, number, number]> = [
  [{ lat: 46.5, lon: 3.0 }, 700_000.0, 6_600_000.0],
  [{ lat: 48.853, lon: 2.3499 }, 652_296.973, 6_861_636.359],
  [{ lat: 45.8452, lon: 6.9051 }, 1_002_953.837, 6_534_776.102],
  [{ lat: 48.3904, lon: -4.4861 }, 146_632.979, 6_836_262.327],
  [{ lat: 43.7102, lon: 7.262 }, 1_043_410.16, 6_299_400.043],
]

describe('Lambert93', () => {
  it('forward projection matches reference coordinates within half a metre', () => {
    for (const [point, x, y] of references) {
      const got = forwardPoint(point)
      expect(Math.abs(got.x - x), `x for ${JSON.stringify(point)}`).toBeLessThan(0.5)
      expect(Math.abs(got.y - y), `y for ${JSON.stringify(point)}`).toBeLessThan(0.5)
    }
  })

  it('inverse round-trips to within a centimetre on the ground', () => {
    for (const [point] of references) {
      const back = inversePoint(forwardPoint(point))
      expect(Math.abs(back.lat - point.lat)).toBeLessThan(1e-7)
      expect(Math.abs(back.lon - point.lon)).toBeLessThan(1e-7)
    }
  })

  it('projection origin lands on the false easting and northing', () => {
    const p = forward(46.5, 3.0)
    expect(p.x).toBeCloseTo(700_000.0, 6)
    expect(p.y).toBeCloseTo(6_600_000.0, 6)
  })

  it('bounds cover the sampled points', () => {
    const b = boundsOf(references.map(([p]) => forwardPoint(p)))
    expect(b.minX).toBeLessThan(200_000)
    expect(b.maxX).toBeGreaterThan(1_000_000)
    expect(boundsContain(expandBounds(b, 1_000_000.0), forward(46.5, 3.0))).toBe(true)
  })

  it('refuses to bound an empty point set', () => {
    expect(() => boundsOf([])).toThrow(/empty/)
  })
})

File: core/src/test/kotlin/io/gpxtoign/core/geo/Lambert93Test.kt
package io.gpxtoign.core.geo

import kotlin.math.abs
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class Lambert93Test {

    /** Reference eastings/northings produced by pyproj (EPSG:4326 -> EPSG:2154). */
    private val references = listOf(
        Triple(LatLon(46.5, 3.0), 700_000.000, 6_600_000.000),
        Triple(LatLon(48.8530, 2.3499), 652_296.973, 6_861_636.359),
        Triple(LatLon(45.8452, 6.9051), 1_002_953.837, 6_534_776.102),
        Triple(LatLon(48.3904, -4.4861), 146_632.979, 6_836_262.327),
        Triple(LatLon(43.7102, 7.2620), 1_043_410.160, 6_299_400.043),
    )

    @Test
    fun `forward projection matches reference coordinates within half a metre`() {
        for ((point, x, y) in references) {
            val got = Lambert93.forward(point)
            assertTrue(abs(got.x - x) < 0.5, "x for $point: got ${got.x}, want $x")
            assertTrue(abs(got.y - y) < 0.5, "y for $point: got ${got.y}, want $y")
        }
    }

    @Test
    fun `inverse round-trips to within a centimetre on the ground`() {
        for ((point, _, _) in references) {
            val back = Lambert93.inverse(Lambert93.forward(point))
            assertTrue(abs(back.lat - point.lat) < 1e-7, "lat for $point: got ${back.lat}")
            assertTrue(abs(back.lon - point.lon) < 1e-7, "lon for $point: got ${back.lon}")
        }
    }

    @Test
    fun `projection origin lands on the false easting and northing`() {
        val p = Lambert93.forward(46.5, 3.0)
        assertEquals(700_000.0, p.x, 1e-6)
        assertEquals(6_600_000.0, p.y, 1e-6)
    }

    @Test
    fun `bounds cover the sampled points`() {
        val b = Bounds.of(references.map { Lambert93.forward(it.first) })
        assertTrue(b.minX < 200_000 && b.maxX > 1_000_000)
        assertTrue(Lambert93.forward(46.5, 3.0) in b.expand(1_000_000.0))
    }
}
